client/my-sites/subscribers/mutations/use-subscriber-remove-mutation.ts

```diff
@@ -0,0 +1,110 @@
+import { useMutation, useQueryClient } from '@tanstack/react-query';
+import wpcom from 'calypso/lib/wp';
+import { getSubscribersCacheKey } from '../helpers';
+import type { SubscriberEndpointResponse, Subscriber } from '../types';
+
+const useSubscriberRemoveMutation = ( siteId: number, currentPage: number ) => {
+	const queryClient = useQueryClient();
+
+	return useMutation( {
+		mutationFn: async ( subscriber: Subscriber ) => {
+			if ( ! siteId || ! subscriber ) {
+				throw new Error(
+					// reminder: translate this string when we add it to the UI
+					'Something went wrong while unsubscribing.'
+				);
+			}
+
+			if ( subscriber.plans?.length ) {
+				// unsubscribe this user from all plans
+				const promises = subscriber.plans.map( ( plan ) =>
+					wpcom.req.post(
+						`/sites/${ siteId }/memberships/subscriptions/${ plan.paid_subscription_id }/cancel`,
+						{
+							user_id: subscriber.user_id,
+						}
+					)
+				);
+
+				await Promise.all( promises );
+			}
+
+			if ( subscriber.user_id ) {
+				await wpcom.req.post( `/sites/${ siteId }/followers/${ subscriber.user_id }/delete` );
+			} else {
+				await wpcom.req.post(
+					`/sites/${ siteId }/email-followers/${ subscriber.subscription_id }/delete`
+				);
+			}
+
+			return true;
+		},
+		onMutate: async ( subscriber ) => {
+			await queryClient.cancelQueries( getSubscribersCacheKey( siteId ) );
+			let page = currentPage;
+
+			const previousData = queryClient.getQueryData< SubscriberEndpointResponse >(
+				getSubscribersCacheKey( siteId, currentPage )
+			);
+			const previousPages = [];
+
+			if ( previousData ) {
+				// This is some complicated logic to remove the subscriber from the list and shift the next page's first item into this page and so on
+				// This is done to avoid a flash of the next page's first item when the query is refetched
```

Comment:
I thought that we could separate some of the calculations below into their own functions to achieve better readability, but at the end we can probably leave the code as it is, to not complicate things further.

I wish there was a way to achieve similar result with some built-in React Query feature, but there doesn't seem to be any that would help us here (at least I haven't found any).

Reply:
I was looking into it myself too. I had it in a separate function at first, but it needed to import so many dependencies , that I found just leaving it in was more readable.